src/Config/MergeConfigProvider.cs

```diff
@@ -0,0 +1,129 @@
+        /// This helper method tries to merge two Json Object recursively.
+        /// </summary>
+        private static void MergeObjects(Utf8JsonWriter jsonWriter, JsonElement root1, JsonElement root2)
+        {
+            jsonWriter.WriteStartObject();
+
+            // If a property exists in both documents, either:
+            // * Merge Json Objects recursively,
+            // * Json Arrays are not merged, but replaced,
+            // * Completely override the value of the first with the one from the second, if the value kind mismatches (e.g. one is object, while the other is an array or string),
+            // * Or favor the value of the first, if the second one is null.
+            foreach (JsonProperty property in root1.EnumerateObject())
+            {
+                string propertyName = property.Name;
+
+                JsonValueKind newValueKind;
+
+                if (root2.TryGetProperty(propertyName, out JsonElement newValue) && (newValueKind = newValue.ValueKind) != JsonValueKind.Null)
+                {
+                    jsonWriter.WritePropertyName(propertyName);
+
+                    JsonElement originalValue = property.Value;
+                    JsonValueKind originalValueKind = originalValue.ValueKind;
+
+                    if (newValueKind == JsonValueKind.Object && originalValueKind == JsonValueKind.Object)
+                    {
+                        MergeObjects(jsonWriter, originalValue, newValue); // Recursive call
+                    }
+                    else if (newValueKind == JsonValueKind.Array && originalValueKind == JsonValueKind.Array)
+                    {
+                        OverrideArrays(jsonWriter, newValue);
+                    }
+                    else
+                    {
+                        newValue.WriteTo(jsonWriter);
+                    }
+                }
+                else
+                {
+                    // Writing Values from the first config as it is NULL (i.e., no override available) in the second config.
+                    property.WriteTo(jsonWriter);
+                }
+            }
+
+            // Write all the properties of the second document that are unique to it.
+            foreach (JsonProperty property in root2.EnumerateObject())
+            {
+                if (!root1.TryGetProperty(property.Name, out _))
+                {
+                    property.WriteTo(jsonWriter);
+                }
+            }
+
+            jsonWriter.WriteEndObject();
+        }
+
+        /// <summary>
+        /// This helper methods completely overrides the elements from the first config with that of second.
+        /// For example: For an entity, The permissions defined in the override config would completely replace
+        /// the permissions of the base config for the same entity in the merged config.
+        /// </summary>
+        private static void OverrideArrays(Utf8JsonWriter jsonWriter, JsonElement root2)
```

Comment:
nit: exlpain what these params mean.

Looking at the params I cant seem to see which param corresponds to the "first" config

Reply:
there are no params corresponding to the first config, we don't need them, because the resulting array will only have elements from second config.

Reply:
could you please provide explanation of the params in the method signature?